GVFS/GVFS.Common/GitHubUpgrader.cs

```diff
@@ -38,71 +38,139 @@ public partial class ProductUpgrader
+            ITracer tracer,
+            out bool isEnabled,
+            out bool isConfigured)
+        {
+            GitHubUpgrader upgrader = null;
+            LocalGVFSConfig localConfig = new LocalGVFSConfig();
+            GitHubUpgraderConfig gitHubUpgraderConfig = new GitHubUpgraderConfig(tracer, localConfig);
+
+            if (gitHubUpgraderConfig.TryLoad(out isEnabled, out isConfigured, out string error))
+            {
+                upgrader = new GitHubUpgrader(
+                    ProcessHelper.GetCurrentProcessVersion(),
+                    tracer,
+                    gitHubUpgraderConfig);
+            }
+
+            return upgrader;
+        }
+
+        public bool Initialize(out string errorMessage)
+        {
+            errorMessage = null;
+            return true;
+        }
+
+        public bool CanRunUsingCurrentConfig(
+            out bool isConfigError, 
+            out string consoleMessage,
+            out string errorMessage)
         {
-            // The values here should be ascending.
-            // Invalid - User has set an incorrect ring
-            // NoConfig - User has Not set any ring yet
-            // None - User has set a valid "None" ring
-            // (Fast should be greater than Slow, 
-            //  Slow should be greater than None, None greater than Invalid.)
-            // This is required for the correct implementation of Ring based 
-            // upgrade logic.
-            Invalid = 0,
-            NoConfig = None - 1,
-            None = 10,
-            Slow = None + 1,
-            Fast = Slow + 1,
+            if (this.Config.UpgradeRing == GitHubUpgraderConfig.RingType.None)
+            {
+                isConfigError = false;
+                consoleMessage = GVFSConstants.UpgradeVerbMessages.NoneRingConsoleAlert + Environment.NewLine + GVFSConstants.UpgradeVerbMessages.SetUpgradeRingCommand;
+                errorMessage = null;
+                return false;
+            }
+
+            if (this.Config.UpgradeRing == GitHubUpgraderConfig.RingType.NoConfig)
+            {
+                isConfigError = false;
+                consoleMessage = GVFSConstants.UpgradeVerbMessages.NoRingConfigConsoleAlert + Environment.NewLine + GVFSConstants.UpgradeVerbMessages.SetUpgradeRingCommand;
+                errorMessage = null;
+                return false;
+            }
+
+            if (this.Config.UpgradeRing == GitHubUpgraderConfig.RingType.Invalid)
+            {
+                string ring;
+                string error;
+                if (!this.Config.LocalConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, out ring, out error))
+                {
+                    ring = "invalid";
+                }
+
+                consoleMessage = null;
+                errorMessage = $"Invalid upgrade ring `{ring}` specified in gvfs config." + Environment.NewLine + GVFSConstants.UpgradeVerbMessages.SetUpgradeRingCommand;

```

Comment:
There are cases where this could read "Invalid upgrade ring `invalid` specified in gvfs config". Let's not use "gvfs" here, it's the product name: VFS for GIt. Also, I don't like "invalid ring invalid". Can we improve that error case?